src/Altinn.App.Core/AppImplementationFactory.cs

```diff
@@ -0,0 +1,57 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Altinn.App.Core.Features;
+
+/// <summary>
+/// Marker attribute for interfaces that are meant to be implemented by apps.
+/// </summary>
+[AttributeUsage(AttributeTargets.Interface, AllowMultiple = false)]
+internal sealed class ImplementableByAppsAttribute : Attribute { }
+
+internal static class AppImplementationFactoryExtensions
+{
+    public static IServiceCollection AddAppImplementationFactory(this IServiceCollection services)
+    {
+        services.AddSingleton<AppImplementationFactory.DIScopeHolder>(sp =>
+            new(sp.GetRequiredService<IHttpContextAccessor>().HttpContext?.RequestServices)
+        );
+        services.AddSingleton<AppImplementationFactory>();
+        return services;
+    }
+
+    public static IServiceCollection AddTestAppImplementationFactory(this IServiceCollection services)
+    {
+        services.AddSingleton<AppImplementationFactory.DIScopeHolder>(sp => new(sp));
+        services.AddSingleton<AppImplementationFactory>();
+        return services;
+    }
+}
+
+internal sealed class AppImplementationFactory
+{
+    internal sealed record DIScopeHolder(IServiceProvider? ScopedServiceProvider);
+
+    private readonly DIScopeHolder _diScopeHolder;
```

Comment:
I find this concept and the additional indirection a little hard to understand.

The straight forward solution would be to just inject a scoped ServiceProvider and use reflection to access the `IsRootScope` to verify when running in debug mode (as we do in tests, but not when apps are running)

```C#
#if DEBUG
        var serviceProviderType = serviceProvider.GetType();
        if (
            serviceProviderType.FullName
            != "Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope"
        )
        {
            throw new InvalidOperationException(
                "InterfaceFactory expects ServiceProviderEngineScope to run debug scope validation"
            );
        }
        var rootProperty = serviceProviderType.GetProperty(
            "IsRootScope",
            System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance
        );
        if (rootProperty == null)
        {
            throw new InvalidOperationException(
                "InterfaceFactory expects ServiceProviderEngineScope to have IsRootScope property"
            );
        }
        var rootPropertyValue = rootProperty.GetValue(serviceProvider);
        switch (rootPropertyValue)
        {
            case true:
                throw new InvalidOperationException("InterfaceFactory should not be used in root scope");
            case false:
                break;
            default:
                throw new InvalidOperationException("InterfaceFactory expects IsRootScope to be a boolean");
        }
#endif
```

Reply:
Hmmm so my only intention with this thing was to enable HTTP-less test fixtures, and that's it. Might have given it a poor name 😄 So ideally what we go with only changes behavior for those specific cases (simple tests using only a DI container). Since for all other cases (app runtime, integration tests) we want the `IHttpContextAccessor` request scope, I'm not sure if the code you posted is justified. To me it seems complicating rather than simplifying

Reply:
I think having a constructor `(IServiceProivder sp)` is simpler (especially for tests) than having to wrap it in an `internal DIScopeHolder`. The code I posted is just part of an internal verification, and not the how the rest of the code base uses the interface factory.

Reply:
The tests don't need to care about that though. There are the 2 DI methods for registering the factory as a service, so in that sense I would say this code is internal to the factory. So the two "modes" of execution for the current code is

1. `AddAppImplementationFactory` -> `IHttpContextAccessor.HttpContext.RequestServices` (for real/app and integration test use)
2. `AddTestAppImplementationFactory` -> Root container (for simple tests where we don't want to spin up ASP.NET Core test host and whatnot)

I added a commit where I get rid of the so called "scope holder" and just use a bool in the constructor instead. So in the first mode we will get an exception if there is no request scope, and in the test mode I don't think we really care